include/glow/Base/Tensor.h

```diff
@@ -48,6 +48,43 @@ void genericTranspose(const Tensor *src, Tensor *dest,
 /// returned dims. For example, input {2,1,4} would result in {2,1,4,1,1,1}.
 ShapeVector expandDimsToMax(llvm::ArrayRef<size_t> currDims);
 
+namespace runtime {
+class DeviceManager;
+}
+
+class DeviceResidencyInfo final {
+  enum class TensorResidency {
+    Host,
+    Device,
+  };
+
+  /// The residency status of the tensor.
+  TensorResidency tensorResidency_{TensorResidency::Host};
+
+  void clear() {
+    deviceManager = nullptr;
+    context = nullptr;
+    tensorResidency_ = TensorResidency::Host;
+  }
+
+public:
+  DeviceResidencyInfo()
+      : tensorResidency_(TensorResidency::Host), deviceManager(nullptr),
+        context(nullptr) {}
+
```

Comment:
I think you should have a destructor here which calls deviceManager->releaseTensor if the tensor is resident.

Reply:
Yes!
What we want to do is actually **give DM an option to release / put-back-in-the-pool** the memory buffer. 

I think we should change the API to releaseDeviceTensor(void *context). 
What's needed in order to free the buffer is the context. Tensor may be used to get the context from DRI. If we want DRI DTOR to call "release" it can pass the stored context.
Is there another motivation to have Tensor& as the input arg?